// @babbage/sdk-ts/out/src/abortAction.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.abortAction = abortAction;
const connectToSubstrate_1 = __importDefault(require("./utils/connectToSubstrate"));
const stampLog_1 = require("./utils/stampLog");
/** Aborts a previously created action which required custom input unlocking script signing.
 * @param {object} args All parameters for this function are provided in an object
 * @param {referenceNumber} args.referenceNumber The referenceNumber associated with a prior createAction result where signActionRequired is true.
 * @returns {Promise<AbortActionResult>} An Action object containing "txid", "rawTx" "mapiResponses" and "inputs".
 */
async function abortAction(args) {
    const connection = await (0, connectToSubstrate_1.default)();
    let log = (0, stampLog_1.stampLog)('', 'start sdk-ts abortAction');
    const r = await connection.dispatch({
        name: 'abortAction',
        params: {
            referenceNumber: args.referenceNumber,
            log
        },
        bodyJsonParams: true,
    });
    r.log = (0, stampLog_1.stampLog)(r.log, 'end sdk-ts abortAction');
    if (typeof args.log === 'string')
        r.log = args.log + r.log;
    return r;
}
exports.default = abortAction;
