// @babbage/sdk-ts/out/src/createCertificate.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCertificate = createCertificate;
const connectToSubstrate_1 = __importDefault(require("./utils/connectToSubstrate"));
/**
 * Creates a signed certificate
 * @param {Object} args All parameters for this function are provided in an object
 * @param {string} args.certificateType The type of certificate to create
 * @param {Object} args.fieldObject The fields to add to the certificate
 * @param {string} args.certifierUrl The URL of the certifier signing the certificate
 * @param {string} args.certifierPublicKey The public identity key of the certifier signing the certificate
 * @returns {Promise<CreateCertificateResult>} A signed certificate
 */
async function createCertificate(args) {
    const connection = await (0, connectToSubstrate_1.default)();
    const r = await connection.dispatch({
        name: 'createCertificate',
        params: {
            certificateType: args.certificateType,
            fieldObject: args.fieldObject,
            certifierUrl: args.certifierUrl,
            certifierPublicKey: args.certifierPublicKey
        },
        bodyJsonParams: true,
    });
    return r;
}
exports.default = createCertificate;
