// @babbage/sdk-ts/out/src/discoverByIdentityKey.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.discoverByIdentityKey = discoverByIdentityKey;
const connectToSubstrate_1 = __importDefault(require("./utils/connectToSubstrate"));
/**
 * Resolves identity information by identity key from the user's trusted certifiers.
 * @param {Object} obj All parameters are provided in an object
 * @param {String} obj.identityKey The identity key to resolve information for
 * @param {string} [obj.description] Describe the high-level operation being performed, so that the user can make an informed decision if permission is needed.
 * @returns {Promise<Object[]>}
 */
async function discoverByIdentityKey(args) {
    const connection = await (0, connectToSubstrate_1.default)();
    const r = await connection.dispatch({
        name: 'discoverByIdentityKey',
        params: {
            identityKey: args.identityKey,
            description: args.description
        },
        isGet: true
    });
    return r;
}
exports.default = discoverByIdentityKey;
