// @babbage/sdk-ts/out/src/getCertificates.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCertificates = getCertificates;
const connectToSubstrate_1 = __importDefault(require("./utils/connectToSubstrate"));
/**
 * Returns found certificates
 * @param {Object} obj All parameters for this function are provided in an object
 * @param {Array<string>} [obj.certifiers] The certifiers to filter certificates by
 * @param {Object} [obj.types] The certificate types to filter certificates by,
 * given as an object whose keys are types
 * and whose values are arrays of fields to request from certificates of the given type.
 * @returns {Promise<CreateCertificateResult[]>} An object containing the found certificates
 */
async function getCertificates(args) {
    const connection = await (0, connectToSubstrate_1.default)();
    const r = await connection.dispatch({
        name: 'findCertificates',
        params: {
            certifiers: args.certifiers,
            types: args.types
        },
        bodyJsonParams: true,
        isNinja: true
    });
    return r;
}
exports.default = getCertificates;
