// @babbage/sdk-ts/out/src/proveCertificate.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.proveCertificate = proveCertificate;
const connectToSubstrate_1 = __importDefault(require("./utils/connectToSubstrate"));
/**
 * Creates certificate proof specifically for verifier
 * @param {Object} args All parameters for this function are provided in an object
 * @param {CertificateApi} args.certificate The certificate to prove
 * @param {Array<string>} args.fieldsToReveal The names of the fields to reveal to the verifier
 * @param {string} args.verifierPublicIdentityKey The public identity key of the verifier
 * @returns {Promise<ProveCertificateResult>} A certificate for presentation to the verifier for field examination
 */
async function proveCertificate(args) {
    const connection = await (0, connectToSubstrate_1.default)();
    const r = await connection.dispatch({
        name: 'proveCertificate',
        params: {
            certificate: args.certificate,
            fieldsToReveal: args.fieldsToReveal,
            verifierPublicIdentityKey: args.verifierPublicIdentityKey
        },
        bodyJsonParams: true,
    });
    return r;
}
exports.default = proveCertificate;
