// @babbage/sdk-ts/out/src/getPreferredCurrency.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPreferredCurrency = getPreferredCurrency;
const connectToSubstrate_1 = __importDefault(require("./utils/connectToSubstrate"));
/**
 * Returns the user's preferred currency for displaying within apps
 *
 * @param {Object} args All parameters are passed in an object.
 * @param {string} [args.description] Describe the high-level operation being performed, so that the user can make an informed decision if permission is needed.
* @returns {Promise<string>} The user's preferred currency
 */
async function getPreferredCurrency(args) {
    const connection = await (0, connectToSubstrate_1.default)();
    const r = await connection.dispatch({
        name: 'getPreferredCurrency',
        nameHttp: 'getPreferredCurrency',
        params: {
            description: args.description
        },
        isGet: true
    });
    return r;
}
exports.default = getPreferredCurrency;
