// @babbage/sdk-ts/out/src/unbasketOutput.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.unbasketOutput = unbasketOutput;
const connectToSubstrate_1 = __importDefault(require("./utils/connectToSubstrate"));
/**
 * Removes the uniquely identified output's basket assignment.
 *
 * The output will no longer belong to any basket.
 *
 * This is typically only useful for outputs that are no longer usefull.
 *
 * @param {Object} args All parameters are given in an object
 * @param {String} [args.txid] The transaction hash as hex string of the transaction containing the output.
 * @param {number} [args.vout] The output index in the transaction.
 * @param {String} [args.basket] The name of the basket from which to remove the output.
 */
async function unbasketOutput(args) {
    const connection = await (0, connectToSubstrate_1.default)();
    await connection.dispatch({
        name: 'unbasketOutput',
        params: {
            txid: args.txid,
            vout: args.vout,
            basket: args.basket
        },
        bodyJsonParams: true,
        isNinja: true
    });
}
exports.default = unbasketOutput;
