// @babbage/sdk-ts/out/src/sdk/ProtoWallet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProtoWallet = void 0;
const KeyDeriver_1 = require("./KeyDeriver");
const WalletCrypto_1 = require("./WalletCrypto");
/**
 * A ProtoWallet is a structure that fulfills the Wallet interface, capable of performing all foundational cryptographic operations. It can derive keys, create signatures, facilitate encryption and HMAC operations, and reveal key linkages. However, ProtoWallet does not create transactions, manage outputs, interact with the blockchain, enable the management of identity certificates, or store any data.
 */
class ProtoWallet extends WalletCrypto_1.WalletCrypto {
    constructor(rootKeyOrKeyDeriver) {
        if (typeof rootKeyOrKeyDeriver['identityKey'] !== 'string') {
            rootKeyOrKeyDeriver = new KeyDeriver_1.KeyDeriver(rootKeyOrKeyDeriver);
        }
        super(rootKeyOrKeyDeriver);
    }
    async isAuthenticated(args, Originator) {
        return { authenticated: true };
    }
    async waitForAuthentication(args, Originator) {
        return { authenticated: true };
    }
    async getNetwork(args, Originator) {
        return { network: 'mainnet' };
    }
    async getVersion(args, Originator) {
        return { version: 'proto-1.0.0' };
    }
    async createAction(args, Originator) {
        throw new Error('ProtoWallet does not support creating transactions.');
    }
    async signAction(args, Originator) {
        throw new Error('ProtoWallet does not support creating transactions.');
    }
    async abortAction(args, Originator) {
        throw new Error('ProtoWallet does not support aborting transactions.');
    }
    async listActions(args, Originator) {
        throw new Error('ProtoWallet does not support retrieving transactions.');
    }
    async internalizeAction(args, Originator) {
        throw new Error('ProtoWallet does not support internalizing transactions.');
    }
    async listOutputs(args, Originator) {
        throw new Error('ProtoWallet does not support retrieving outputs.');
    }
    async relinquishOutput(args, Originator) {
        throw new Error('ProtoWallet does not support deleting outputs.');
    }
    async acquireCertificate(args, Originator) {
        throw new Error('ProtoWallet does not support acquiring certificates.');
    }
    async listCertificates(args, Originator) {
        throw new Error('ProtoWallet does not support retrieving certificates.');
    }
    async proveCertificate(args, Originator) {
        throw new Error('ProtoWallet does not support proving certificates.');
    }
    async relinquishCertificate(args, Originator) {
        throw new Error('ProtoWallet does not support deleting certificates.');
    }
    async discoverByIdentityKey(args, Originator) {
        throw new Error('ProtoWallet does not support resolving identities.');
    }
    async discoverByAttributes(args, Originator) {
        throw new Error('ProtoWallet does not support resolving identities.');
    }
    async getHeight(args, Originator) {
        throw new Error('ProtoWallet does not support blockchain tracking.');
    }
    async getHeaderForHeight(args, Originator) {
        throw new Error('ProtoWallet does not support blockchain tracking.');
    }
}
exports.ProtoWallet = ProtoWallet;
